crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/Libc.hs

```diff
@@ -782,6 +825,86 @@ callBSwap ::
   OverrideSim p sym ext r args ret (RegValue sym (BVType (width * 8)))
 callBSwap sym widthRepr (regValue -> vec) = liftIO $ bvSwap sym widthRepr vec
 
+-- | This determines under what circumstances @callAbs@ should check if its
+-- argument is equal to the smallest signed integer of a particular size
+-- (e.g., @INT_MIN@), and if it is equal to that value, what kind of error
+-- should be reported.
+data CheckAbsIntMin
+  = LibcAbsIntMinUB
+    -- ^ For the @abs@, @labs@, and @llabs@ functions, always check if the
+    --   argument is equal to @INT_MIN@. If so, report it as undefined
+    --   behavior per the C standard.
+  | LLVMAbsIntMinPoison Bool
+    -- ^ For the @llvm.abs.*@ family of LLVM intrinsics, check if the argument
+    --   is equal to @INT_MIN@ only when the 'Bool' argument is 'True'. If it
+    --   is 'True' and the argument is equal to @INT_MIN@, return poison.
+
+-- | The workhorse for the @abs@, @labs@, and @llabs@ functions, as well as the
+-- @llvm.abs.*@ family of overloaded intrinsics.
+callAbs ::
+  forall w p sym ext r args ret.
+  (1 <= w, IsSymInterface sym, HasLLVMAnn sym) =>
+  sym ->
+  CheckAbsIntMin ->
+  NatRepr w ->
+  RegEntry sym (BVType w) ->
+  OverrideSim p sym ext r args ret (RegValue sym (BVType w))
+callAbs sym checkIntMin widthRepr (regValue -> src) = liftIO $ do
+  bvIntMin      <- bvLit sym widthRepr (BV.minSigned widthRepr)
+  isNotIntMin   <- notPred sym =<< bvEq sym src bvIntMin
+  isSrcNegative <- bvIsNeg sym src
+  srcNegated    <- bvNeg sym src
+  srcAbs        <- bvIte sym isSrcNegative srcNegated src
+  if shouldCheckIntMin
+     then do isNotIntMinUB <- annotateUB sym ub isNotIntMin
+             let err = AssertFailureSimError "Undefined behavior encountered" $
+                       show $ UB.explain ub
+             assert sym isNotIntMinUB err
+             pure srcAbs
+     else -- Per https://releases.llvm.org/12.0.0/docs/LangRef.html#id452,
+          -- return INT_MIN if the argument is INT_MIN and we are not returning
+          -- poison.
+          bvIte sym isNotIntMin srcAbs src
```

Comment:
I think we can simplify this.  If `src = INT_MIN` then `negate src = INT_MIN` also, so I think you can just return `srcAbs`.

Reply:
Good catch. I've applied your suggestion.